common/djangoapps/split_modulestore_django/models.py

```diff
@@ -0,0 +1,165 @@
+
+from xmodule.modulestore import ModuleStoreEnum
+
+User = get_user_model()
+
+
+class SplitModulestoreCourseIndex(models.Model):
+    """
+    A "course index" for a course in "split modulestore."
+
+    This model/table mostly stores the current version of each course.
+    (Well, twice for each course - "draft" and "published" branch versions are
+    tracked separately.)
+
+    This MySQL table / django model is designed to replace the "active_versions"
+    MongoDB collection. They contain the same information.
+
+    It also stores the "wiki_slug" to facilitate looking up a course
+    by it's wiki slug, which is required due to the nuances of the
+    django-wiki integration.
+
+    .. no_pii:
+    """
+    # For compatbility with MongoDB, each course index must have an ObjectId. We still have an integer primary key too.
+    objectid = models.CharField(max_length=24, null=False, blank=False, unique=True)
+
+    # The ID of this course (or library). Must start with "course-v1:" or "library-v1:"
+    course_id = LearningContextKeyField(max_length=255, db_index=True, unique=True, null=False)
+    # Extract the "org" value from the course_id key so that we can search by org.
+    # This gets set automatically by clean()
+    org = models.CharField(max_length=255, db_index=True)
+
+    # Version fields: The ObjectId of the current entry in the "structures" collection, for this course.
+    # The version is stored separately for each "branch".
+    # Note that there are only three branch names allowed. Draft/published are used for courses, while "library" is used
+    # for content libraries.
+
+    # ModuleStoreEnum.BranchName.draft = 'draft-branch'
+    draft_version = models.CharField(max_length=255, null=False, blank=True)
+    # ModuleStoreEnum.BranchName.published = 'published-branch'
+    published_version = models.CharField(max_length=255, null=False, blank=True)
+    # ModuleStoreEnum.BranchName.library = 'library'
+    library_version = models.CharField(max_length=255, null=False, blank=True)
+
+    # Wiki slug for this course
+    wiki_slug = models.CharField(max_length=255, db_index=True, blank=True)
+
+    # Base store - whether the "structures" and "definitions" data are in MongoDB or object storage (S3)
+    BASE_STORE_MONGO = "mongodb"
+    BASE_STORE_DJANGO = "django"
+    BASE_STORE_CHOICES = [
+        (BASE_STORE_MONGO, "MongoDB"),  # For now, MongoDB is the only implemented option
+        (BASE_STORE_DJANGO, "Django - not implemented yet"),
+    ]
+    base_store = models.CharField(max_length=20, blank=False, choices=BASE_STORE_CHOICES)
+
+    # Edit history:
+    # ID of the user that made the latest edit. This is not a ForeignKey because some values (like
+    # ModuleStoreEnum.UserID.*) are not real user IDs.
+    edited_by_id = models.IntegerField(null=True)
+    edited_on = models.DateTimeField()
+    # last_update is different from edited_on, and is used only to prevent collisions?
```

Comment:
> It'd be nice to have some clarity here on the usage of this field. Is it only used to prevent collisions?

It is only used in the `mongo_connection` file, and not even referenced in the main `split.py` implementation, so yes, that does seem to be its only use. But to be honest I'm not totally clear on the reason it exists or how it works. For now I just tried to copy the existing behavior exactly.

The PR that introduced it is here, does this help at all? https://github.com/edx/edx-platform/pull/5200

I will look into it a bit more myself after I investigate something else, see below.

> It seems wrong here to not log anything? At the very least, we'd want to know the frequency at which this situation occurs, since an update is being ignored. In what situation would this situation be expected to happen?

The existing code [uses `last_update` as a filter on the MongoDB replace_one method query](https://github.com/edx/edx-platform/blob/58d61734790524be1e8b6ef073ae67c23f60188e/common/lib/xmodule/xmodule/modulestore/split_mongo/mongo_connection.py#L450-L453), which I believe has the effect of silently ignoring updates if the previous `last_update` version doesn't match. So this is intended to preserve that behavior. I'm not sure either of what circumstances that would be expected to happen, so perhaps logging is a good idea.

In fact, I added some logging locally and did a test in Studio where I mashed the "Add Video" button eight times in rapid succession, and although I didn't discover any `last_update` "collisions", I unfortunately found that there is some issue with this code that results in conflicting writes to the database - fewer than eight videos end up getting added. I will have to look into it a bit more to figure out what's going on. It seems to mostly work on the master branch (eight videos do eventually get added, but there are other errors), so I think there is some transaction-related issue here.

Reply:
Right, it's probably intended to stop some race condition where you have conflicting writes with slow transactions, but I don't know how effective it's been in practice.